lib/src/decimal128.dart

```diff
@@ -0,0 +1,149 @@
+////////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+import 'dart:convert';
+import 'dart:ffi';
+
+import 'package:ffi/ffi.dart';
+
+import 'native/realm_bindings.dart';
+import 'native/realm_core.dart';
+
+/// A 128-bit decimal floating point number.
+class Decimal128 extends Comparable<Decimal128> {
+  /// The value 0.
+  static final zero = Decimal128.fromInt(0);
+
+  /// The value 1.
+  static final one = Decimal128.fromInt(1);
+
+  /// The value 10.
+  static final ten = Decimal128.fromInt(10);
+
+  /// The value NaN.
+  static final nan = Decimal128._(lib.realm_dart_decimal128_nan());
+
+  /// The value +Inf.
+  static final infinity = one / zero; // +Inf
+
+  /// The value -Inf.
+  static final negativeInfinity = -infinity;
+
+  final realm_decimal128_t _value;
+
+  Decimal128._(this._value);
+
+  static final _validInput = RegExp(r'^[+-]?((\d+\.?\d*|\d*\.?\d+)([eE][+-]?\d+)?|NaN|Inf(inity)?)$');
+
+  /// Parses a string into a [Decimal128]. Returns `null` if the string is not a valid [Decimal128].
+  static Decimal128? tryParse(String source) {
+    if (!_validInput.hasMatch(source)) return null;
+    return using((arena) {
+      final result = lib.realm_dart_decimal128_from_string(source.toNativeUtf8(allocator: arena).cast());
+      return Decimal128._(result);
+    });
+  }
+
+  /// Parses a string into a [Decimal128]. Throws a [FormatException] if the string is not a valid [Decimal128].
+  factory Decimal128.parse(String source) {
+    return tryParse(source) ?? (throw FormatException('Invalid Decimal128', source));
+  }
+
+  /// Converts a `int` into a [Decimal128].
+  factory Decimal128.fromInt(int value) {
+    return Decimal128._(lib.realm_dart_decimal128_from_int64(value));
+  }
+
+  /// Converts a `double` into a [Decimal128].
+  factory Decimal128.fromDouble(double value) {
+    return Decimal128.parse(value.toString()); // TODO(kn): Find a way to optimize this
```

Comment:
One TODO comment has left here.

Reply:
Yeah.. it is a bit of a bummer.. Let me look it over again

Reply:
We got to live with this for now :-/